sdks/python/apache_beam/runners/portability/universal_local_runner.py

```diff
@@ -0,0 +1,398 @@
+    super(UniversalLocalRunner, self).__init__()
+    self._use_grpc = use_grpc
+    self._use_subprocesses = use_subprocesses
+
+    self._handle = None
+    self._subprocess = None
+
+  def __del__(self):
+    # Best effort to not leave any dangling processes around.
+    self.cleanup()
+
+  def cleanup(self):
+    if self._subprocess:
+      self._subprocess.kill()
+      time.sleep(0.1)
+    self._subprocess = None
+
+  def _get_handle(self):
+    if not self._handle:
+      if self._use_subprocesses:
+        if self._subprocess:
+          # Kill the old one if it exists.
+          self._subprocess.kill()
+        # TODO(robertwb): Consider letting the subprocess pick one and
+        # communicate it back...
+        port = _pick_unused_port()
+        logging.info("Starting server on port %d.", port)
+        self._subprocess = subprocess.Popen([
+            sys.executable,
+            '-m',
+            'apache_beam.runners.portability.universal_local_runner_main',
+            '-p',
+            str(port),
+            '--worker_command_line',
+            '%s -m apache_beam.runners.worker.sdk_worker_main' % sys.executable
+        ])
+        handle = beam_job_api_pb2_grpc.JobServiceStub(
+            grpc.insecure_channel('localhost:%d' % port))
+        logging.info("Waiting for server to be ready...")
+        start = time.time()
+        timeout = 30
+        while True:
+          time.sleep(0.1)
+          if self._subprocess.poll() is not None:
+            raise RuntimeError(
+                "Subprocess terminated unexpectedly with exit code %d." %
+                self._subprocess.returncode)
+          elif time.time() - start > timeout:
+            raise RuntimeError(
+                "Pipeline timed out waiting for job service subprocess.")
+          else:
+            try:
+              handle.GetState(
+                  beam_job_api_pb2.GetJobStateRequest(job_id='[fake]'))
+              break
+            except grpc.RpcError as exn:
+              if exn.code != grpc.StatusCode.UNAVAILABLE:
+                break
+        logging.info("Server ready.")
+        self._handle = handle
+
+      elif self._use_grpc:
+        self._servicer = JobServicer(use_grpc=True)
+        self._handle = beam_job_api_pb2_grpc.JobServiceStub(
+            grpc.insecure_channel('localhost:%d' % self._servicer.start_grpc()))
+
+      else:
+        self._handle = JobServicer(use_grpc=False)
+
+    return self._handle
+
+  def run(self, pipeline):
+    handle = self._get_handle()
+    prepare_response = handle.Prepare(
+        beam_job_api_pb2.PrepareJobRequest(
+            job_name='job',
+            pipeline=pipeline.to_runner_api()))
+    run_response = handle.Run(beam_job_api_pb2.RunJobRequest(
+        preparation_id=prepare_response.preparation_id))
+    return PipelineResult(handle, run_response.job_id)
+
+
+class PipelineResult(runner.PipelineResult):
+  def __init__(self, handle, job_id):
```

Comment:
<!--new_thread; commit:77e51af3ee2af39264c366a2602957e7e9bb20c7; resolved:0-->
Let's document constructor arguments in Docstring. For example, it might not be obvious what handle is in this context.

Reply:
Renamed handle.